src/main/java/org/springframework/data/couchbase/repository/query/StringBasedN1qlQueryParser.java

```diff
@@ -177,50 +182,64 @@ private String getDistinctFields(String... distinctFields) {
+	private String getProjectedFields(String b, Class resultClass, String[] fields) {
 
 		String projectedFields = i(b) + ".*";
 		if (resultClass != null) {
 			PersistentEntity persistentEntity = couchbaseConverter.getMappingContext().getPersistentEntity(resultClass);
 			StringBuilder sb = new StringBuilder();
-			getProjectedFieldsInternal(b, null, sb, persistentEntity.getTypeInformation()/*, ""*/);
+			getProjectedFieldsInternal(b, null, sb, persistentEntity.getTypeInformation(), fields/*, ""*/);
 			projectedFields = sb.toString();
 		}
 		return projectedFields;
 	}
 
 	private void getProjectedFieldsInternal(String bucketName, CouchbasePersistentProperty parent, StringBuilder sb,
-			TypeInformation resultClass/*, String path*/) {
-
-		PersistentEntity persistentEntity = couchbaseConverter.getMappingContext().getPersistentEntity(resultClass);
-		// CouchbasePersistentProperty property = path.getLeafProperty();
-		persistentEntity.doWithProperties(new PropertyHandler<CouchbasePersistentProperty>() {
-			@Override
-			public void doWithPersistentProperty(final CouchbasePersistentProperty prop) {
-				if (prop.isIdProperty() && parent == null) {
-					return;
-				}
-				if (prop.isVersionProperty()) {
-					return;
-				}
-				PersistentPropertyPath<CouchbasePersistentProperty> path = couchbaseConverter.getMappingContext()
-						.getPersistentPropertyPath(prop.getName(), resultClass.getType());
-
-				// The current limitation is that only top-level properties can be projected
-				// This traversing of nested data structures would need to replicate the processing done by
-				// MappingCouchbaseConverter. Either the read or write
-				// And the n1ql to project lower-level properties is complex
+			TypeInformation resultClass, String[] fields/*, String path*/) {
 
-				// if (!conversions.isSimpleType(prop.getType())) {
-				// getProjectedFieldsInternal(prop, sb, prop.getTypeInformation(), path+prop.getName()+".");
-				// } else {
+		if (resultClass != null) {
+			Set<String> fieldsList = fields != null ? new HashSet<>(Arrays.asList(fields)) : null;
+			PersistentEntity persistentEntity = couchbaseConverter.getMappingContext().getPersistentEntity(resultClass);
+			// CouchbasePersistentProperty property = path.getLeafProperty();
+			persistentEntity.doWithProperties(new PropertyHandler<CouchbasePersistentProperty>() {
+				@Override
+				public void doWithPersistentProperty(final CouchbasePersistentProperty prop) {
+					if (prop.isIdProperty() && parent == null) {
+						return;
+					}
+					if (prop.isVersionProperty()) {
+						return;
+					}
+					if (fieldsList != null && !fieldsList.contains(prop.getFieldName())) {
+						return;
+					}
+					PersistentPropertyPath<CouchbasePersistentProperty> path = couchbaseConverter.getMappingContext()
+							.getPersistentPropertyPath(prop.getName(), resultClass.getType());
+
+					// The current limitation is that only top-level properties can be projected
```

Comment:
can we detect this and throw an error if so?

Reply:
Yes.  Fixed.